packages/core/src/notifications/controller.ts

```diff
@@ -0,0 +1,233 @@
+        this.notificationsNode = node
+
+        this.state = globals.globalState.get(this.storageKey) ?? {
+            startUp: {} as NotificationData,
+            emergency: {} as NotificationData,
+            dismissed: [],
+        }
+        this.state.startUp = this.state.startUp ?? {}
+        this.state.emergency = this.state.emergency ?? {}
+        this.state.dismissed = this.state.dismissed ?? []
+    }
+
+    public pollForStartUp(ruleEngine: RuleEngine) {
+        return this.poll(ruleEngine, 'startUp')
+    }
+
+    public pollForEmergencies(ruleEngine: RuleEngine) {
+        return this.poll(ruleEngine, 'emergency')
+    }
+
+    private async poll(ruleEngine: RuleEngine, category: NotificationType) {
+        try {
+            await this.fetchNotifications(category)
+        } catch (err: any) {
+            getLogger().error(`Unable to fetch %s notifications: %s`, category, err)
+        }
+
+        await this.displayNotifications(ruleEngine)
+    }
+
+    private async displayNotifications(ruleEngine: RuleEngine) {
+        const dismissed = new Set(this.state.dismissed)
+        const startUp =
+            this.state.startUp.payload?.notifications.filter(
+                (n) => !dismissed.has(n.id) && ruleEngine.shouldDisplayNotification(n)
+            ) ?? []
+        const emergency = (this.state.emergency.payload?.notifications ?? []).filter((n) =>
+            ruleEngine.shouldDisplayNotification(n)
+        )
+
+        NotificationsNode.instance.setNotifications(startUp, emergency)
+
+        // Emergency notifications can't be dismissed, but if the user minimizes the panel then
+        // we don't want to focus it each time we set the notification nodes.
+        // So we store it in dismissed once a focus has been fired for it.
+        const newEmergencies = emergency.map((n) => n.id).filter((id) => !dismissed.has(id))
+        if (newEmergencies.length > 0) {
+            this.state.dismissed = [...this.state.dismissed, ...newEmergencies]
+            await this.writeState()
+            void this.notificationsNode.focusPanel()
+        }
+    }
+
+    /**
+     * Permanently hides a notification from view. Only 'startUp' notifications can be dismissed.
+     * Users are able to collapse or hide the notifications panel in native VSC if they want to
+     * hide all notifications.
+     */
+    public async dismissNotification(notificationId: string) {
+        getLogger().debug('Dismissing notification: %s', notificationId)
+        this.state.dismissed.push(notificationId)
+        await this.writeState()
+
+        NotificationsNode.instance.dismissStartUpNotification(notificationId)
+    }
+
+    /**
+     * Fetch notifications from the endpoint and store them in the global state.
+     */
+    private async fetchNotifications(category: NotificationType) {
+        const response = _useLocalFiles ? await this.fetchLocally(category) : await this.fetchRemotely(category)
```

Comment:
Another option is to use dependency injection and pass an object that matches `fetchX`, then in the class constructor you could pass the instance you want for tests.

But I'm assuming you don't have access to the constructor at runtime, so maybe creating a top level function in this file where you can set a fetch implementation (otherwise defaulting to the http one), would work best. Then the constructor grabs whatever is there at instantiation.

This will get rid of the custom test environment logic

Reply:
Its not meant for use in test cases - only during manual testing when you have the extension running in debug mode or you are testing a packaged vsix.

Reply:
I updated it anyway, the new flow looks good and does help the unit test case path as a side effect.